### ISSUE TITLE
[Question] Is there a way to call delegating constructor?

### ISSUE THREAD

**Issue by thedmd** [hsutter/cppfront#409]
I'm plaing with cppfront trying to port some bits of code to get a feel of new syntax.

One thing I cannot figure out is how to write `operator=:` to delegate construction.
I tried `operator=: (out this, major: u32) = this(major, 0, 0, 0);`, `operator=: (out this, major: u32) = this := (major, 0, 0, 0);`, even stared with `operator=: (out this, major: u32) = Version(major, 0, 0, 0);`.

In cppfront source I didn't found code emitting, but I'm not yet 100% sure.

Did I missed the intended syntax or feature is not present at this time? Thanks!

For reference:
```
Version: @value type =
{
    Major:      u32 = 0;
    Minor:      u32 = 0;
    Revision:   u32 = 0;
    Patch:      u32 = 0;

    operator=: (out this, major: u32)                                        = { this.Major = major; }
    operator=: (out this, major: u32, minor: u32)                            = { this.Major = major; this.Minor = minor; }
    operator=: (out this, major: u32, minor: u32, revision: u32)             = { this.Major = major; this.Minor = minor; this.Revision = revision; }
    operator=: (out this, major: u32, minor: u32, revision: u32, patch: u32) =
    {
        this.Major    = major;
        this.Minor    = minor;
        this.Revision = revision;
        this.Patch    = patch;
    }

```

**Comment by JohelEGP** [comment]
It seems there's no way.

IIRC, you should also be able to delegate to any `out` parameter: <https://cpp2.godbolt.org/z/TGWP6fKWY>.
```Cpp2
f: (out x: t) = { x = 0; }
t: type = {
  a: i32;
  operator=: (out this, b: i32) = { a = b; }
  operator=: (out this, c: std::string) = {
    f(this);
  }
}
main: () -> int = t("").a;
```
```output
main.cpp2(6,5): error: expected 'a = ...' initialization statement - an operator= body must start with a series of 'member = value;' initialization statements for each of the type-scope objects in the same order they are declared
```

**Comment by realgdman** [comment]
Well calling directly seems to work

```
Version: @value type = {
	Major: u32 = 0;
	Minor: u32 = 0;
	operator=: (out this, major: u32) = Major = major;
	operator=: (out this, major: u32, minor: u32) = { 
		operator=(major); 
		Minor = minor; 
	}
	print: (this) = std::cout << "(Major)$.(Minor)$\n";
}

main: () = {
	v1: Version = (11);
	v2: Version = (20, 23);
	
	v1.print(); //11.0
	v2.print(); //20.23
}
```
However technically thats not delegated constructor, but calling assign operator (which I think provided according to that diagram from [wiki]( https://github.com/hsutter/cppfront/wiki/Cpp2:-operator=,-this-&-that)

**Comment by JohelEGP** [comment]
Interesting. I tried with explicit `this`, but it didn't occur to me to omit it.

**Comment by realgdman** [comment]
Btw is there syntax for initialization like cpp1 `Ctor(a, b): a(a), b(b)` ?

**Comment by JohelEGP** [comment]
No. Just use assignments. See https://github.com/hsutter/cppfront/wiki/Cpp2:-operator=,-this-&-that#misc-notes.

**Comment by realgdman** [comment]
Probably there is no need for delegating constructors at all and this can be solved in the future with just one constructor, default values, and named parameters in.
Like, `v:= Version(.major = 5, .minor = 3)`

**Comment by JohelEGP** [comment]
I think it's an accident that <https://github.com/hsutter/cppfront/issues/409#issuecomment-1533063425> works.
If you remove `Major`'s member initializer, it still breaks.
That's why <https://github.com/hsutter/cppfront/issues/409#issuecomment-1533030507> doesn't wok with `operator=(0);`.

**Comment by realgdman** [comment]
Well I think it works but not exactly how it is expected.
Ideally it should automatically detect that initialization happens in all code paths.
So now it expecting straightforward either in declaration `i : int = 0` or inside constructor `i = 0`
But in case of calling operator= it is not tracking that member was initialized inside,so initializing in delcaration suppress error, but constructor overwrites value later.
I think there is some overhead but it is allowed to work. It's like this code
```
S: type = {
	i: int = 2;
	operator=: (out this) = { i = 3; }
	print: (this) = std::cout << i;
}
```
Which boils to `i=2; i=3;`

**Comment by JohelEGP** [comment]
> Which boils down to `i=2; i=3;`

Looking closely, that (double assignment) seems to be the case
for the `Version` code that explicitly calls `operator=`.

**Comment by realgdman** [comment]
In wiki it said that cpp2 guarantees initialization, that means initializing at least once, but it's not clear if that means exactly once and not more.

**Comment by JohelEGP** [comment]
> (Also I don't know c++ well, may be its bad to call member function before constructor is finished...)

It's OK to.

**Comment by JohelEGP** [comment]
I think assignment to `this` should delegate.

Also, something should be decided about `operator=`.
In Cpp2,
> `operator=` is the name for all value-setting functions
> -- https://github.com/hsutter/cppfront/wiki/Cpp2:-operator=,-this-&-that

But currently, it lowers to Cpp1 as `operator=`.

**Comment by thedmd** [comment]
Thanks for exploring the issue.

My understanding is that delegating constructors are not supported right now, but can be introduced in the future. Is that correct?

---

Few thoughts that came to me while reading the thread:

> https://github.com/hsutter/cppfront/issues/409#issuecomment-1533075240

Cpp2 encourage use of overloading while default values for arguments are not supported. Default arguments would fit nicely in the scenario with `Version` type.

> https://github.com/hsutter/cppfront/issues/409#issuecomment-1533063425

Calling directly does work, but also reverse logic where most general constructors chain calls to more specialized. It feel more like workaround due to initialization code being more convoluted in the end.

> https://github.com/hsutter/cppfront/issues/409#issuecomment-1533130465

Double assignment here is ok'is. Feel like compromise.
With complex types involved this can easily became no-fly zone.
